bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadOnlyLedgerHandle.java

```diff
@@ -207,4 +206,190 @@ public void readLastConfirmedComplete(int rc, long lastConfirmed, Object ctx) {
+                 final boolean forceRecovery) {
+        final GenericCallback<Void> cb = new TimedGenericCallback<Void>(
+            finalCb,
+            BKException.Code.OK,
+            clientCtx.getClientStats().getRecoverOpLogger());
+
+        MetadataUpdateLoop.NeedsUpdatePredicate needsUpdate =
+            (metadata) -> !(metadata.isClosed() || metadata.isInRecovery());
+        if (forceRecovery) {
+            // in the force recovery case, we want to update the metadata
+            // to IN_RECOVERY, even if the ledger is already closed
+            needsUpdate = (metadata) -> !metadata.isInRecovery();
+        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                needsUpdate,
+                (metadata) -> LedgerMetadataBuilder.from(metadata).withInRecoveryState().build(),
+                this::setLedgerMetadata)
+            .run()
+            .thenCompose((metadata) -> {
+                    if (metadata.isClosed()) {
+                        return CompletableFuture.<Void>completedFuture(null);
+                    } else {
+                        return new LedgerRecoveryOp(ReadOnlyLedgerHandle.this, clientCtx)
+                            .setEntryListener(listener)
+                            .initiate();
+                    }
+            })
+            .thenCompose((ignore) -> closeRecovered())
+            .whenComplete((ignore, ex) -> {
+                    if (ex != null) {
+                        cb.operationComplete(getExceptionCode(ex, BKException.Code.UnexpectedConditionException), null);
+                    } else {
+                        cb.operationComplete(BKException.Code.OK, null);
+                    }
+            });
+    }
+
+    CompletableFuture<LedgerMetadata> closeRecovered() {
+        long lac, len;
+        synchronized (this) {
+            lac = lastAddConfirmed;
+            len = length;
+        }
+        LOG.info("Closing recovered ledger {} at entry {}", getId(), lac);
+        CompletableFuture<LedgerMetadata> f = new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> metadata.isInRecovery(),
+                (metadata) -> {
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    Optional<Long> lastEnsembleKey = metadata.getLastEnsembleKey();
+                    checkState(lastEnsembleKey.isPresent(),
+                               "Metadata shouldn't have been created without at least one ensemble");
+                    newEnsemblesFromRecovery.entrySet().forEach(
+                            (e) -> {
+                                // Occurs when a bookie need to be replaced at very start of recovery
+                                if (lastEnsembleKey.get().equals(e.getKey())) {
```

Comment:
Do we have a specific test case?

Reply:
Ya, I want to test this more, hence the mock reg client (and some more).